releasing/releasing/releasing.go

```diff
@@ -0,0 +1,478 @@
+		if len(args) != 2 {
+			return errors.New("2 arguments are required")
+		}
+		found := false
+		for _, mod := range modules {
+			if mod == args[0] {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("%s is not a valid module. Valid modules are %s", args[0], modules)
+		}
+		types := []string{"major", "minor", "patch"}
+		found = false
+		for _, t := range types {
+			if t == args[1] {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("%s is not a valid version type. Valid types are %s", args[1], types)
+		}
+		return nil
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		modName := args[0]
+		versionType := args[1]
+		createTempDir()
+		logInfo("Creating tag for module %s", modName)
+		pwd, err := os.Getwd()
+		if err != nil {
+			logFatal(err.Error())
+		}
+		logDebug("Working directory: %s", pwd)
+		remote := "upstream"
+		// Check remotes
+		checkRemoteExistence(pwd, remote)
+		// Fetch latest tags from remote
+		fetchTags(pwd, remote)
+
+		mod := module{
+			name: modName,
+			path: pwd,
+		}
+		mod.UpdateCurrentVersion()
+
+		oldVersion := mod.version.String()
+		mod.version.Bump(versionType)
+		newVersion := mod.version.String()
+		logInfo("Bumping version: %s => %s", oldVersion, newVersion)
+
+		// Create branch
+		branch := fmt.Sprintf("release-%s-v%d.%d", mod.name, mod.version.major, mod.version.minor)
+		newBranch(pwd, branch)
+
+		addWorktree(pwd, tempDir, branch)
+
+		merge(tempDir, "upstream/master")
+		// Update module path
+		mod.path = tempDir
+
+		logInfo(
+			"Releasing summary:\nDir:\t%s\nModule:\t%s %s\nBranch:\t%s\nTag:\t%s",
+			tempDir,
+			mod.name,
+			mod.version.String(),
+			branch,
+			mod.Tag(),
+		)
+
+		// Run module tests
+		output, err := mod.RunTest()
+		if err != nil {
+			logWarn(output)
+		} else if !noDryRun {
+			logInfo("Skipping push module %s. Run with --no-dry-run to push the release.", mod.name)
+		} else {
+			pushRelease(tempDir, branch, mod)
+		}
+		// Clean
+		removeTempDir()
+		pruneWorktree(pwd)
+		deleteBranch(pwd, branch)
+		logInfo("Releasing for module %s completes", mod.name)
+	},
+}
+
+var subCmds = [...]*cobra.Command{
+	listSubCmd,
+	release,
+}
+
+// === Main function ===
+
+func main() {
+	for _, cmd := range subCmds {
+		rootCmd.AddCommand(cmd)
+	}
+	rootCmd.PersistentFlags().BoolVarP(&verbose, "verbose", "v", false, "verbose output")
+	release.Flags().BoolVarP(&noDryRun, "no-dry-run", "", false, "disable dry-run")
+	release.Flags().BoolVarP(&doTest, "do-test", "", false, "run module tests before releasing")
+
+	if err := rootCmd.Execute(); err != nil {
+		logFatal(err.Error())
+	}
+}
+
+func getModuleCurrentVersion(modName, path string) string {
+	mod := module{
+		name: modName,
+		path: path,
+	}
+	mod.UpdateCurrentVersion()
+	v := mod.version.String()
+	logDebug("module %s version.toString => %s", mod.name, v)
+	return v
+}
+
+// === module version struct and functions definition ===
+
+type moduleVersion struct {
```

Comment:
please move this to `moduleversion.go` so as we grow it we can add tests to `moduleversion_test.go`